Publish on the blog (#17)
Zenika/labs-tooling: ci-cd-platform-deployment/README.md

# CI/CD Platform Deployment

As people start new projects, they usually waste precious time deploying and configuring a CI/CD pipeline from scratch.
At Zenika Labs, our goal is to deliver proofs of concept or minimum viable products as efficiently as possible, without
compromising on quality.

This guide describes how to setup, in probably less than an hour, the *feature branch* workflow we use everyday to 
build, test and deploy our applications. It uses Google's managed Kubernetes (GKE) as well as CircleCI, Traefik and 

> others Google Cloud services.

Reviewer: "other Google Cloud services" (sans le s)

**What is Kubernetes?**
> Kubernetes, also known as K8s, is an open-source system for automating deployment, scaling, and management of 
containerized applications. In the past years, Kubernetes has become the de-facto industrial standard to deploy containers on-premise or in the
cloud.
 
If you have never used Kubernetes before, this guide is probably not for you. Start by 
[learning a bit about Kubernetes](https://kubernetes.io/docs/concepts/overview/) first.

**What is CircleCI?**
> CircleCI is a cloud-native continuous integration and continuous delivery (CI/CD) platform. It integrates with GitHub
and Bitbucket and runs a configured pipeline on each commit. Think Jenkins multibranch pipelines on steroids, in the 
cloud, and fully managed for you.
 
If you have never used CircleCI before, welcome aboard and enjoy the [free plan](https://circleci.com/pricing/)!

**What is Traefik?**

> > Traefik is a dynamic, modern, open-source Edge Router that automatically inspects your infrastructure to discover 

Reviewer 2: Est-ce que ça vaudrait pas le coup d'utiliser le nouveau nom "Traefik Proxy" ?  🤔

services and how to route traffic to them. Traefik is natively compliant with every major cluster technology, such as 
Kubernetes, Docker, Docker Swarm, AWS, Mesos, Marathon... and even bare metal! Used as an ingress provider in 

> Kubernetes, it is probably a drop-in replacement for the default provider and brings awesome features such as automated

Reviewer 2: Je ne comprends pas le fait d'être un remplacement du "default provider" 🤔  De base il n'y a pas d'Ingress Controller dans Kubernetes. Dans GKE il  y a bien un Ingress Controller par défaut par contre.

Author: You're right c'était pas clair. J'ai reformulé.

At Zenika Labs, each of our projects has its own namespace in a shared, autoscaling Kubernetes cluster. Every new 
feature is implemented in a new, short-lived branch. As shown later with the [sample project](sample_project), our 
CI/CD pipeline allows us to create and maintain an application instance per feature branch, accessible from anywhere 
(web) and at anytime. This scheme goes in the direction of our *Definition of Done*, which requires a code review and a
validation from a Product Owner on a production-like environment before deploying to production.

The following guide contains a lot of data and commands gathered from the following resources:
- [Kubernetes' Quickstart](https://cloud.google.com/kubernetes-engine/docs/quickstart)
- [Traefik's official documentation](https://docs.traefik.io)

Read them in details if you want to deep dive into how things work.

## Architecture and outline

Our pipeline uses GitHub for version control, CircleCI for CI/CD, a Google Kubernetes Engine cluster as the deployment
platform, Traefik as ingress controller and load balancer and Let's Encrypt for a fully automated and free TLS 

> certificates management. The corresponding architecture is pictured below:

Reviewer: "TLS certificate management" (sans le s)

can be generated through the [Cloud Console](https://console.cloud.google.com/iam-admin/serviceaccounts). 

Note that if you opt for another provider, you will probably need to adapt or remove the `volume`, `volumeMount` and 
environment variables parts of the [traefik descriptor](traefik.yaml) in order to pass the correct configuration to 
Traefik.

Import the key file into your cluster's traefik namespace as a secret with key `traefik-service-account`:
```shell script
kubectl create secret generic traefik-service-account --from-file=traefik-service-account.json=[FILENAME] --namespace=traefik
```

Fill the following placeholders in the [traefik descriptor](traefik.yaml): 
- `DOMAIN`: the domain you own (e.g., `mywebsite.com`)
- `ACME_EMAIL_ADDRESS`: the contact email address to use to generate the TLS certificates
- `GCE_PROJECT`: the name of the Google Cloud project

Then apply it:
```shell script
kubectl apply -f traefik.yaml
```

This will:
- instantiate a Traefik instance using a Deployment
- expose this Traefik instance on a public IP using a Service of type LoadBalancer

> - configure Traefik's entrypoints to listen to ports 80 (http) and 443 (https)

Reviewer 2: ```suggestion
- configure Traefik's entrypoints to listen to ports 80 (HTTP) and 443 (HTTPS)
```

> - redirect all http (port 80) traffic to the https entrypoint (port 443) using a RedirectScheme middleware

Reviewer 2: ```suggestion
- redirect all HTTP (port 80) traffic to the HTTPS entrypoint (port 443) using a RedirectScheme middleware
```

- expose Traefik's dashboard on the `traefik` subdomain (e.g., `traefik.mywebsite.com`) using an IngressRoute, protected
with a BasicAuth middleware (using the secret created above)
- configure a Traefik certificate resolver to generate wildcard certificates on demand
- create and use the wildcard TLS certificate (e.g., `*.mywebsite.com`) required by the dashboard IngressRoute 

> Wait for a bit and get the public IP associated by GKE to the Traefik service:

Reviewer 2: ```suggestion
Wait for a bit and get the public IP associated by GKE to the Traefik Service:
```

```shell script
kubectl -n traefik get services
```
The IP will eventually be displayed in the "EXTERNAL-IP" column, but it may take a few seconds.

Configure your DNS records manually to redirect all traffic from your domain to this IP (this is an `A` record from 
`*.mywebsite.com` to the external IP).

> Traefik's dashboard should now be accessible on the traefik subdomain (e.g., `traefik.mywebsite.com`) and all http

Reviewer 2: ```suggestion
Traefik's dashboard should now be accessible on the traefik subdomain (e.g., `traefik.mywebsite.com`) and all HTTP
```

> traffic should be redirected to https with valid Let's Encrypt certificates.

Reviewer 2: ```suggestion
traffic should be redirected to HTTPS with valid Let's Encrypt certificates.
```

> 

Reviewer 2: Eventuellement je ferai plusieurs articles, il y a déjà beaucoup d'informations. Ici serait une bonne première étape, qu'en penses-tu ? 😃

Author: Bonne idée !

## 3. A first "whoami" toy project

In this section, we will illustrate how to use Traefik to expose multiple instances/branches/versions of a same app on 
subdomains. Suppose we chose to dedicate the `whoami` subdomain (e.g., `whoami.mywebsite.com`) to this project. Our goal
is to expose the master version of the app on a `master` subdomain (e.g., `master.whoami.mywebsite.com`) and the changes
of a `feature1` branch on the `feature1` subdomain (e.g., `feature1.whoami.mywebsite.com`).

For this example, we will use the containous/whoami docker image as a dummy web application, which only serves some 
information about the server and the received request for any request on port 80. 

### Create a namespace for the project

```shell script
kubectl create namespace whoami
```

### Deploy the master version

Fill the following placeholders in the [whoami-master descriptor](whoami-master.yaml): 
- `DOMAIN`: the domain you own (e.g., `mywebsite.com`)

Then apply it:
```shell script
kubectl apply -f whoami-master.yaml
```

> This will instantiate 3 instances of the whoami pod, create a service to load balance between them, and expose this

Reviewer 2: ```suggestion
This will instantiate 3 instances of the whoami Pod, create a Service to load balance between them, and expose this
```

> service through an IngressRoute. Notice the use of the `traefik` TLS certificate resolver created in step 2, the 

Reviewer 2: ```suggestion
Service through an IngressRoute. Notice the use of the `traefik` TLS certificate resolver created in step 2, the 
```

request for a `*.whoami.[DOMAIN]` wildcard certificate and the use of the `master.whoami.[DOMAIN]` Host rule to route

> traffic to this URL to our service. In just a few seconds, the certificate is generated and your application server is

Reviewer 2: ```suggestion
traffic to this URL to our Service. In just a few seconds, the certificate is generated and your application server is
```

exposed securely on the `master.whoami` of your domain (try it!). 

### Deploy the feature1 branch

Fill the following placeholders in the [whoami-branch1 descriptor](whoami-branch1.yaml): 
- `DOMAIN`: the domain you own (e.g., `mywebsite.com`)

Then apply it:
```shell script
kubectl apply -f whoami-branch1.yaml
```

> This will instantiate 3 new instances of the pod, create a new service to load balance between them, and expose this

Reviewer 2: ```suggestion
This will instantiate 3 new instances of the Pod, create a new Service to load balance between them, and expose this
```

> service through an IngressRoute. Notice again the use of the `traefik` TLS certificate resolver created in step 2, the 

Reviewer 2: ```suggestion
Service through an IngressRoute. Notice again the use of the `traefik` TLS certificate resolver created in step 2, the 
```

use of the `branch1.whoami.[DOMAIN]` Host rule to route traffic to this URL to our service. Your new app instance,
distinct from the master environment is already live (try it!). 

### Use a template

Notice the very few differences between the [whoami-master](whoami-master.yaml) and the 
[whoami-branch1](whoami-branch1.yaml) descriptors: all occurences of `master` have simply be replaced by `branch1`. In 
real world use cases, the docker image will probably be different too.

Take a look at the [whoami-template descriptor template](whoami-template.yaml) file. Env-like placeholders (e.g., 
`${INSTANCE_NAME}`) are used. We will use the `envsubst` tool to instantiate our template.

Fill the following placeholders in the [whoami-template descriptor template](whoami-template.yaml): 
- `DOMAIN`: the domain you own (e.g., `mywebsite.com`)
Do not touch the env-like placeholders.

Set the required environement variables, then instantiate the template and apply it: 
```shell script
export INSTANCE_NAME=branch2
export IMAGE=containous/whoami:v1.5.0
envsubst < whoami-template.yaml > whoami-branch2.yaml
kubectl apply -f whoami-branch2.yaml
```

Quick, your `branch2` instance is already up!

> The `envsubst` approach will get you quite far, but if you need a more advanced templating engine (and much more), look

Reviewer 2: ```suggestion
The `envsubst` approach will get you quite far, but if you need a more advanced configuration overlays (and much more), look
```

> into [Kustomize](https://github.com/kubernetes-sigs/kustomize) or [Helm](https://helm.sh/).

Reviewer 2: ```suggestion
into [Kustomize](https://github.com/kubernetes-sigs/kustomize) or [Helm](https://helm.sh/) if you want a more advanced templating engine.
```

> 

Reviewer 2: Fin du 2ème article ?

Author: Il risque d'être un peu léger le deuxième article non ? Peut-être juste partie 1 et partie 2 du coup ?

Reviewer 2: Effectivement 😅

## 4. A "myapp" project with CI/CD

In this section, we will illustrate how the manual deployment described in step 3 can be integrated in about any CI/CD
pipeline. Suppose we chose to dedicate the `myapp` subdomain (e.g., `myapp.mywebsite.com`) to this project. Our goal
is to expose any branch X on the X subdomain (e.g., branch `feat1` exposed on `feat1.myapp.mywebsite.com`, `feat2`
exposed on `feat2.myapp.mywebsite.com`, etc.).

> For this example, we will build our own docker image in a CircleCI pipeline. For the sake of simplicity, this guide will

Reviewer 2: ```suggestion
For this example, we will build our own Docker image in a CircleCI pipeline. For the sake of simplicity, this guide will
```

### Build the project on CircleCI

In CircleCI, add your project in the build list in the `Add Projects` tab (use the `Add manually` button, since you have
already committed a `.circleci/config.yml` file).

In the `Settings` tab, find your project in the `Projects` sub-tab and open its settings.

In `Environment Variables`, create the following variables. They are used in the CircleCI configuration and in the
descriptor template.
- `GCP_PROJECT`: The id of your GCP project (visible in the 
[Cloud Console](https://console.cloud.google.com/home/))
- `GCP_AUTH`: The content of the JSON key file created in the previous step.
- `GCP_REGISTRY`: The GCR registry of your choice (`eu.gcr.io/[GCP_PROJECT]` for a storage in Europe)
- `GCP_ZONE`: The GCP Zone in which your cluster is hosted (visible in the 
[Cloud Console](https://console.cloud.google.com/kubernetes/list), e.g., `europe-west3-a`)
- `GCP_KUBE_CLUSTER`: The name of your Kubernetes cluster

Relaunch the build: the master branch is deployed on the `master` subdomain (e.g., `master.myapp.mywebsite.com`).

Create branches on your GitHub repository and push them... and let the magic happen! From source code to production in 

> less than 20 secondes!

Reviewer: "seconds" (sans le e)
